fish_completions.go

```diff
@@ -43,16 +43,13 @@ function __%[1]s_perform_completion
     end
     __%[1]s_debug "emptyArg: $emptyArg"
 
-    if not type -q "$args[1]"
-        # This can happen when "complete --do-complete %[2]s" is called when running this script.
```

Comment:
Removing this seems sensible. I'm not sure if the comment really matters, because the this fish completion script should only be sourced when the command exists.

Reply:
This scenario is pretty obscure actually.  Let me explain.  

Fish comes with existing completion scripts.  For instance, it has its own `helm` completion script:
https://github.com/fish-shell/fish-shell/blob/master/share/completions/helm.fish
This is script is out of date however and provides some completions that are no longer valid for `helm`.

In the Cobra script this PR updates, we do `complete -c helm -e` to erase all other completions for `helm`.  The problem is that the old `helm` script included in fish may not yet be loaded, which means the user could get completions from both scripts (the old one provided by fish, and the Cobra one).  To fix this, the Cobra script does `complete --do-complete "helm " > /dev/null 2>&1` which forces the loading of the old completion script, so that all these completions can be erased by `complete -c helm -e`.

**The problem happens for programs that are not on PATH**.  In this case, if the Cobra completion script is sourced more than once, calling for example, `complete --do-complete "myprog " > /dev/null 2>&1` would cause fish to print an error because it cannot find `myprog`, and this error printout could not be redirected to `/dev/null`.  To fix that I was catching this special case with `if not type -q "$args[1]"` above.

However, with @Luap99's help, we have found a way to redirect such error to `/dev/null` (see line 52 below).  That is why we no longer need to catch this special case and can remove this code.

Reply:
> This is script is out of date however and provides some completions that are no longer valid for helm.

So the fish script should be changed to a single line `helm completion fish | source`?
Ok this is https://github.com/fish-shell/fish-shell/issues/7357, no one has done that yet. I'll try to do that, I just need to detect the Helm version.

> In the Cobra script this PR updates, we do complete -c helm -e to erase all other completions for helm.

In an ideal world this would not be necessary, because fish's version is installed to, say `/usr/share/fish/completions/helm.fish` but the one that ships with the Helm package is in `/usr/share/fish/vendor_completions.d/helm.fish`, which overrides the former because it's earlier in `$fish_complete_path`.

But for manual (and repeated) sourcing we do need this, so you are right.

I feel like the best solution is to check at the call site if the program is executable:

```fish
if type -q "%[2]s"
    complete --do-complete "%[2]s " > /dev/null 2>&1
end
```

Starting a new fish instance to be able to redirect the error seems a bit overkill.
fish always sources its startup files, so the new instance might cause obscure issues; better to use `eval` (as suggested below).

Reply:
> So the fish script should be changed to a single line `helm completion fish | source`?
> Ok this is [fish-shell/fish-shell#7357](https://github.com/fish-shell/fish-shell/issues/7357), no one has done that yet. I'll try to do that, I just need to detect the Helm version.

You can probably prefix the existing script with something like
```
if helm completion -h | grep -q fish > /dev/null
    helm completion fish | source
else
   # Existing old script
end
```
Note that `helm completion fish` will return a success code even if the command does not exist (#1156), that is why I suggest using the help text to figure out if fish is supported.

> In an ideal world this would not be necessary, because fish's version is installed to, say `/usr/share/fish/completions/helm.fish` but the one that ships with the Helm package is in `/usr/share/fish/vendor_completions.d/helm.fish`, which overrides the former because it's earlier in `$fish_complete_path`.

You are right.  We could choose not to support the explicit sourcing.  In fact, Cobra does not support it for `zsh`.  However, I do find it nicer to support it, as long as the side-effects can be easily controlled.
 
> But for manual (and repeated) sourcing we do need this, so you are right.

👍 

> I feel like the best solution is to check at the call site if the program is executable:
> 
> ```fish
> if type -q "%[2]s"
>     complete --do-complete "%[2]s " > /dev/null 2>&1
> end
> ```

I think that would work, and it is better.  However we thought about adding the extra fish instance to solve another issue where an error was being thrown when a quote was not properly closed, see #1214 

> Starting a new fish instance to be able to redirect the error seems a bit overkill.
> fish always sources its startup files, so the new instance might cause obscure issues; better to use `eval` (as suggested below).

We currently use eval, but it prints error message we cannot catch in the case of #1214

Reply:
> However, I do find explicit sourcing nicer

Yeah, I think it's fine.

> We currently use eval, but it prints error message we cannot catch

The error can be cleanly avoided by escaping `lastArg`, see https://github.com/spf13/cobra/pull/1249#discussion_r563348012.

Reply:
> The error can be cleanly avoided by escaping lastArg, see #1249 (comment).

That works!
So we can do as you suggest:
```
if type -q "%[2]s"
    complete --do-complete "%[2]s " > /dev/null 2>&1
end
```
and avoid having to call a new fish shell to avoid error printouts.

@Luap99 you may be interested in trying the version I will soon push.  But I can confirm that I tested it for unmatched quotes, and completion simply isn't done, with no error message.   Also, the regression tests repo passes as before.